components/brave_rewards/resources/extension/brave_rewards/content.ts

```diff
@@ -0,0 +1,145 @@
+  braveTipIcon.style.display = 'inline-block'
+  braveTipIcon.style.fontSize = '18px'
+  braveTipIcon.style.fontStyle = 'normal'
+  braveTipIcon.style.height = '16px'
+  braveTipIcon.style.marginTop = '5px'
+  braveTipIcon.style.position = 'relative'
+  braveTipIcon.style.verticalAlign = 'baseline'
+  braveTipIcon.style.width = '16px'
+  braveTipIconContainer.appendChild(braveTipIcon)
+
+  // Create the tip action count (typically used to present a counter
+  // associated with the action, but we'll use it to display a static
+  // action label)
+  const braveTipActionCount = document.createElement('span')
+  braveTipActionCount.className = 'ProfileTweet-actionCount'
+  braveTipActionCount.style.color = '#657786'
+  braveTipActionCount.style.display = 'inline-block'
+  braveTipActionCount.style.fontSize = '12px'
+  braveTipActionCount.style.fontWeight = 'bold'
+  braveTipActionCount.style.lineHeight = '1'
+  braveTipActionCount.style.marginLeft = '6px'
+  braveTipActionCount.style.position = 'relative'
+  braveTipActionCount.style.verticalAlign = 'text-bottom'
+  braveTipButton.appendChild(braveTipActionCount)
+
+  // Create the tip action count presentation
+  const braveTipActionCountPresentation = document.createElement('span')
+  braveTipActionCountPresentation.className = 'ProfileTweet-actionCountForPresentation'
+  braveTipActionCountPresentation.textContent = getMessage('twitterTipsIconLabel')
+  braveTipActionCount.appendChild(braveTipActionCountPresentation)
+
+  // Create the shadow DOM root that hosts our injected DOM elements
+  const shadowRoot = braveTipAction.attachShadow({ mode: 'open' })
+  shadowRoot.appendChild(braveTipButton)
+
+  return braveTipAction
+}
+
+const configureBraveTipAction = () => {
+  chrome.runtime.sendMessage('rewardsEnabled', function (rewards) {
+    const msg = {
+      type: 'inlineTipSetting',
+      key: 'twitter'
+    }
+    chrome.runtime.sendMessage(msg, function (inlineTip) {
+      const tweets = document.getElementsByClassName('tweet')
+      for (let i = 0; i < tweets.length; ++i) {
+        const actions = tweets[i].getElementsByClassName('js-actions')[0]
+        if (actions) {
+          const braveTipActions = actions.getElementsByClassName('action-brave-tip')
+          const tippingEnabled = rewards.enabled && inlineTip.enabled
+          if (tippingEnabled && braveTipActions.length === 0) {
+            actions.appendChild(createBraveTipAction(tweets[i]))
+          } else if (!tippingEnabled && braveTipActions.length === 1) {
+            actions.removeChild(braveTipActions[0])
+          }
+        }
+      }
+    })
+  })
+  setTimeout(configureBraveTipAction, 3000)
```

Comment:
We're running this function every 3 seconds for the lifetime of every twitter page? I'm now guessing this how we're able to inject the buttons on new items loaded for infinite scroll / overlays. It would be nice to be able to do this more intelligently, only on these options, but I understand that anything like that would be more difficult and potentially flakey. Still, I wonder if this affects:
- performance (UX jank)
- power usage (iterating through the entire DOM on every twitter.com site every 3 seconds)
At the very least, if we can't only run this when the content changes, perhaps we could only run it when the tab is in the foreground and the content is visible?

Reply:
Yes, this was my attempt to get around the inifinite scroll. I tried a couple of other approaches initially, but wasn't able to find anything that worked very reliably outside of this. Running it only when foregrounded/etc. makes absolute sense, will get that in there at a minimum. Outside of that, I would say let's revisit when we integrate with our site-specific scripting solution.